File: python/rest-api/functions/create.py
import json
import dataclasses
from datetime import datetime
from dataclasses import dataclass, field

from nitric import faas
from nitric.faas import HttpContext, start
from nitric.api import Documents
from common.example import generate_id

@dataclass(frozen=True)
class OrderRequest:
  itemId: str
  customerId: str
  id: str = field(default_factory=lambda: str(generate_id()))
  dateOrdered: str = field(default_factory=lambda: str(datetime.now()))

async def httpHandler(ctx: HttpContext) -> HttpContext:
  order = OrderRequest(**json.loads(ctx.req.data))
  orders = Documents().collection('orders')

  try:
    await orders.doc(order.id).set(dataclasses.asdict(order))
    ctx.res.status = 200
    ctx.res.body = f'Created example with ID: {order.id}'.encode('utf-8')
  except:
    ctx.res.status = 500
    ctx.res.body = b'Failed to create order'

  return ctx

if __name__ == "__main__":
  start(httpHandler)

File: typescript/rest-api/functions/create.ts
// [START snippet]
// functions/create.ts

import { faas, documents } from "@nitric/sdk";
import { uuid } from "uuidv4";
import { Order } from "../common";

// Requests to create a new order won't have an id or order date
type RequestData = Omit<Order, "id" | "dateOrdered">;

interface CreateContext extends faas.HttpContext {
  req: faas.HttpRequest & {
    body?: RequestData;
  };
}

interface CreateResponse {
  message: string;
  orderId?: string;
}

// Start your function here
faas
  .http(
    faas.json(), //  use json body parser middleware to decode data
    async (ctx: CreateContext): Promise<faas.HttpContext> => {
      const data = ctx.req.body;
      const response = ctx.res;

      const order: Order = {
        ...data, // extract data from the request
        id: uuid(), // generate a new uuid
        dateOrdered: new Date().toJSON(),
      };

      const orders = documents().collection<Order>("orders");

      try {
        // Write the new order document to the orders collection
        await orders.doc(order.id).set(order);
        response.body = order.id;
        response.json({
          message: "success",
          orderId: order.id,
        } as CreateResponse);
      } catch {
        response.status = 500;
        response.body = new TextEncoder().encode("failed to create order");
      }

      return ctx;
    }
  )
  .start();

// [END snippet]